docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,426 @@
+
+The main reason for this choice is because a managed topology has a set of capabilities required for lifecycle hooks implementation:
+
+* A managed topology has the overarching view of the entire Cluster (vs other components in CAPI which are limited to controlling single resources or a subset of them).
+* A managed topology already has the capability to control all resources in a Cluster, thus making it possible to orchestrate lifecycle workflows like e.g. upgrades.
+
+Please note that this is especially relevant because it allows treating the cluster lifecycle as a state machine no matter the underlying controller being based on level-triggered reconcile loops.
+
+###  User Stories
+
+These user stories are based on a concrete example of an add-on - a metrics database -  to illustrate the use of each Runtime Hook.
+
+As a developer of an add-ons orchestration solution:
+
+* **Before a Cluster is Created** I want to automatically check if enough disk space is available for allocation to the cluster for persistent storage of collected metrics values.
+* **After the Control Plane** **is Initialized** I want to automatically install a metrics database and associated add-ons in the workload cluster.
+* **Before the Cluster is Upgraded** I want to install a new version of the metrics database with a new version of the custom metrics apiservice to interact directly with the Kubernetes apiserver.
+* **After the ControlPlane is Upgraded** I want to automatically check that the new version of the custom metrics apiservice is working and correctly fulfilled by my metrics database.
+* **After the Cluster is Upgraded** I want to install new versions of metrics collectors to each upgraded node in the cluster.
+* **Before the Cluster is Deleted** I want to automatically back up persistent volumes used by the metrics database.
+
+### Runtime hook definitions
+
+Below is a description for the Runtime Hooks introduced by this proposal.
+
+![runtime-hooks](images/runtime-hooks/runtime-hooks.png)
+
+
+#### Before Cluster Create
+
+This hook is called after the Cluster object has been created by the user, immediately before all the objects which are part of a Cluster topology(*) are going to be created. Runtime Extension implementers can use this hook to determine/prepare add-ons for the Cluster and block the creation of those objects until everything is ready. 
+
+##### Example Request:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterCreateRequest
+cluster:
+  apiVersion: cluster.x-k8s.io/v1beta1
+  kind: Cluster
+  metadata:
+   name: test-cluster
+   namespace: test-ns
+  spec:
+   ...
+  status:
+   ...
+```
+
+##### Example Response:
+
+```yaml
+apiVersion: hooks.runtime.cluster.x-k8s.io/v1alpha1
+kind: BeforeClusterCreateResponse
+status: Success
+message: "error message if status == Failure"
```

Comment:
what's status in this context?

Reply:
The status field in the type can be Success or Failure. If it's marked as success there should be a message adding context in this message field.